Support NJT chunk() backward on batch dim

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #144889
* #144588
* #144587
* #144586
* #144585
* __->__ #144584
* #144583
* #144582

Part of my BE project addressing NJT bugs surfaced via OpInfo tests.

Implements `chunk()` backward on the batch dim, which was left out before. This PR unbinds the components and invokes `copy_()` on these to pass along the appropriate gradients.

jbschlosser committed:
Support NJT chunk() backward on batch dim

[ghstack-poisoned]

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -652,9 +652,19 @@ def copy_default(func, *args, **kwargs):
-        raise RuntimeError(
-            "copy_ only supports Nested Tensors that have same size and the exact same offset tensor."
-        )
+        # try to recursively copy_ on unbound components
```
Cool way to get around the nested int mismatch! 
Should we add a TODO to say that we'd eventually want to directly copy?

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -8343,14 +8343,6 @@ def __torch_dispatch__(self, func, types, args=..., kwargs=None):
-        op_match_fn=lambda device, op: (op.full_name == "chunk"),
-        sample_match_fn=lambda device, sample: ("batch_dim" in sample.name),
-        name="broken_chunk_backward",
-    ),
```
May need to update the `test_copy_` test which asserts `copy_ only supports Nested Tensors that have same size and the exact same offset tensor`

jbschlosser committed:
Update on "Support NJT chunk() backward on batch dim"


Part of my BE project addressing NJT bugs surfaced via OpInfo tests.

Implements `chunk()` backward on the batch dim, which was left out before. This PR unbinds the components and invokes `copy_()` on these to pass along the appropriate gradients.

[ghstack-poisoned]

jbschlosser committed:
Update on "Support NJT chunk() backward on batch dim"


Part of my BE project addressing NJT bugs surfaced via OpInfo tests.

Implements `chunk()` backward on the batch dim, which was left out before. This PR unbinds the components and invokes `copy_()` on these to pass along the appropriate gradients.

[ghstack-poisoned]

jbschlosser committed:
Update on "Support NJT chunk() backward on batch dim"


Part of my BE project addressing NJT bugs surfaced via OpInfo tests.

Implements `chunk()` backward on the batch dim, which was left out before. This PR unbinds the components and invokes `copy_()` on these to pass along the appropriate gradients.

[ghstack-poisoned]

jbschlosser committed:
Update on "Support NJT chunk() backward on batch dim"


Part of my BE project addressing NJT bugs surfaced via OpInfo tests.

Implements `chunk()` backward on the batch dim, which was left out before. This PR unbinds the components and invokes `copy_()` on these to pass along the appropriate gradients.

[ghstack-poisoned]

Skylion007 commented:
@pytorchbot merge
